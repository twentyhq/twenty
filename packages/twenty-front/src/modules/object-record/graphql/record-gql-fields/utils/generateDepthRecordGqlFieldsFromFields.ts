import { type FieldMetadataItem } from '@/object-metadata/types/FieldMetadataItem';
import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { getImageIdentifierFieldMetadataItem } from '@/object-metadata/utils/getImageIdentifierFieldMetadataItem';
import { getLabelIdentifierFieldMetadataItem } from '@/object-metadata/utils/getLabelIdentifierFieldMetadataItem';
import { type RecordGqlFields } from '@/object-record/graphql/record-gql-fields/types/RecordGqlFields';
import { FieldMetadataType, RelationType } from 'twenty-shared/types';
import { computeMorphRelationFieldName, isDefined } from 'twenty-shared/utils';

export type GenerateDepthRecordGqlFieldsFromFields = {
  objectMetadataItems: ObjectMetadataItem[];
  fields: Pick<
    FieldMetadataItem,
    'name' | 'type' | 'settings' | 'morphRelations' | 'relation'
  >[];
  depth: 0 | 1;
  shouldOnlyLoadRelationIdentifiers?: boolean;
};

export const generateDepthRecordGqlFieldsFromFields = ({
  objectMetadataItems,
  fields,
  depth,
  shouldOnlyLoadRelationIdentifiers = true,
}: GenerateDepthRecordGqlFieldsFromFields) => {
  const generatedRecordGqlFields: RecordGqlFields = fields.reduce(
    (recordGqlFields, fieldMetadata) => {
      const isMorphRelation =
        fieldMetadata.type === FieldMetadataType.MORPH_RELATION;

      const isRelation = fieldMetadata.type === FieldMetadataType.RELATION;

      const relationType = fieldMetadata.settings?.relationType;

      if (isRelation) {
        const targetObjectMetadataItem = objectMetadataItems.find(
          (objectMetadataItem) =>
            objectMetadataItem.id ===
            fieldMetadata.relation?.targetObjectMetadata.id,
        );

        if (!targetObjectMetadataItem) {
          throw new Error(
            `Target object metadata item not found for ${fieldMetadata.name}`,
          );
        }

        const labelIdentifierFieldMetadataItem =
          getLabelIdentifierFieldMetadataItem(targetObjectMetadataItem);

        const imageIdentifierFieldMetadataItem =
          getImageIdentifierFieldMetadataItem(targetObjectMetadataItem);

        const relationIdentifierSubGqlFields = {
          id: true,
          ...(isDefined(labelIdentifierFieldMetadataItem)
            ? { [labelIdentifierFieldMetadataItem.name]: true }
            : {}),
          ...(isDefined(imageIdentifierFieldMetadataItem)
            ? { [imageIdentifierFieldMetadataItem.name]: true }
            : {}),
        };

        const manyToOneGqlFields = {
          [`${fieldMetadata.name}Id`]: true,
        };

        return {
          ...recordGqlFields,
          ...(depth === 1 && shouldOnlyLoadRelationIdentifiers
            ? { [fieldMetadata.name]: relationIdentifierSubGqlFields }
            : undefined),
          ...(depth === 1 && !shouldOnlyLoadRelationIdentifiers
            ? { [fieldMetadata.name]: true }
            : undefined),
          ...(relationType === RelationType.MANY_TO_ONE
            ? manyToOneGqlFields
            : undefined),
        };
      }

      if (isMorphRelation) {
        if (!isDefined(fieldMetadata.morphRelations)) {
          throw new Error(
            `Field ${fieldMetadata.name} is missing, please refresh the page. If the problem persists, please contact support.`,
          );
        }

        const morphGqlFields = fieldMetadata.morphRelations.map(
          (morphRelation) => ({
            gqlField: computeMorphRelationFieldName({
              fieldName: fieldMetadata.name,
              relationType: morphRelation.type,
              targetObjectMetadataNameSingular:
                morphRelation.targetObjectMetadata.nameSingular,
              targetObjectMetadataNamePlural:
                morphRelation.targetObjectMetadata.namePlural,
            }),
            fieldMetadata,
          }),
        );

        return {
          ...recordGqlFields,
          ...morphGqlFields.reduce(
            (morphGqlFields, morphGqlField) => ({
              ...morphGqlFields,
              ...(depth === 1
                ? { [`${morphGqlField.gqlField}`]: { id: true, name: true } }
                : {}),
              ...(relationType === RelationType.MANY_TO_ONE
                ? { [`${morphGqlField.gqlField}Id`]: true }
                : {}),
            }),
            {},
          ),
        };
      }

      return {
        ...recordGqlFields,
        [`${fieldMetadata.name}`]: true,
      };
    },
    {},
  );

  return generatedRecordGqlFields;
};
