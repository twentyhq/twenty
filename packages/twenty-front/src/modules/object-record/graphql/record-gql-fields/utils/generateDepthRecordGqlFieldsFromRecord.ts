import { type RecordGqlFields } from '@/object-record/graphql/record-gql-fields/types/RecordGqlFields';
import {
  generateDepthRecordGqlFieldsFromObject,
  type GenerateDepthRecordGqlFields,
} from '@/object-record/graphql/record-gql-fields/utils/generateDepthRecordGqlFieldsFromObject';
import { type ObjectRecord } from '@/object-record/types/ObjectRecord';

type ComputeDepthRecordGqlFieldsFromRecordArgs =
  GenerateDepthRecordGqlFields & {
    record: Partial<ObjectRecord>;
  };
export const generateDepthRecordGqlFieldsFromRecord = ({
  objectMetadataItem,
  objectMetadataItems,
  depth,
  record,
}: ComputeDepthRecordGqlFieldsFromRecordArgs): RecordGqlFields => {
  const depthRecordGqlFields = generateDepthRecordGqlFieldsFromObject({
    objectMetadataItem,
    objectMetadataItems,
    depth,
  });
  const recordKeys = Object.keys(record);

  return Object.keys(depthRecordGqlFields).reduce<RecordGqlFields>(
    (acc, key) => {
      return {
        ...acc,
        [key]: recordKeys.includes(key),
      };
    },
    depthRecordGqlFields,
  );
};
