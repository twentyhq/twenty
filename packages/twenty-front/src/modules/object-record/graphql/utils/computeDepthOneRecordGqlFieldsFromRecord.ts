import {
  GenerateDepthOneRecordGqlFields,
  generateDepthOneRecordGqlFields,
} from '@/object-record/graphql/utils/generateDepthOneRecordGqlFields';
import { ObjectRecord } from '@/object-record/types/ObjectRecord';

type ComputeDepthOneRecordGqlFieldsFromRecordArgs =
  GenerateDepthOneRecordGqlFields & {
    record: Partial<ObjectRecord>;
  };
export const computeDepthOneRecordGqlFieldsFromRecord = ({
  objectMetadataItem,
  record,
}: ComputeDepthOneRecordGqlFieldsFromRecordArgs) => {
  const depthOneRecordGqlFields = generateDepthOneRecordGqlFields({
    objectMetadataItem,
  });
  const recordKeys = Object.keys(record);

  return Object.keys(depthOneRecordGqlFields).reduce<Record<string, boolean>>(
    (acc, key) => {
      return {
        ...acc,
        [key]: recordKeys.includes(key),
      };
    },
    depthOneRecordGqlFields,
  );
};
