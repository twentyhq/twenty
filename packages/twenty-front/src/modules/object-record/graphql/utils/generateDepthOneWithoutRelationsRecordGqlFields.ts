import { ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { FieldMetadataType } from '~/generated/graphql';

type GenerateDepthOneWithoutRelationsRecordGqlFields = {
  objectMetadataItem: ObjectMetadataItem;
};

export const generateDepthOneWithoutRelationsRecordGqlFields = ({
  objectMetadataItem,
}: GenerateDepthOneWithoutRelationsRecordGqlFields) => {
  return objectMetadataItem.fields
    .filter((field) => field.type !== FieldMetadataType.RELATION)
    .reduce<Record<string, true>>((acc, field) => {
      return {
        ...acc,
        [field.name]: true,
      };
    }, {});
};
