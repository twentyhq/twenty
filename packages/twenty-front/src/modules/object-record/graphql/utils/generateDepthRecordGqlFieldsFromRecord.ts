import { type RecordGqlFields } from '@/object-record/graphql/types/RecordGqlFields';
import {
  generateDepthRecordGqlFields,
  type GenerateDepthRecordGqlFields,
} from '@/object-record/graphql/utils/generateDepthRecordGqlFields';
import { type ObjectRecord } from '@/object-record/types/ObjectRecord';

type ComputeDepthRecordGqlFieldsFromRecordArgs =
  GenerateDepthRecordGqlFields & {
    record: Partial<ObjectRecord>;
  };
export const generateDepthRecordGqlFieldsFromRecord = ({
  objectMetadataItem,
  objectMetadataItems,
  depth,
  record,
}: ComputeDepthRecordGqlFieldsFromRecordArgs): RecordGqlFields => {
  const depthRecordGqlFields = generateDepthRecordGqlFields({
    objectMetadataItem,
    objectMetadataItems,
    depth,
  });
  const recordKeys = Object.keys(record);

  return Object.keys(depthRecordGqlFields).reduce<RecordGqlFields>(
    (acc, key) => {
      return {
        ...acc,
        [key]: recordKeys.includes(key),
      };
    },
    depthRecordGqlFields,
  );
};
