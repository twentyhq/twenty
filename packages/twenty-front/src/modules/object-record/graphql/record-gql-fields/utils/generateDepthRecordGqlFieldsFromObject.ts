import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { generateDepthRecordGqlFieldsFromFields } from '@/object-record/graphql/record-gql-fields/utils/generateDepthRecordGqlFieldsFromFields';

export type GenerateDepthRecordGqlFields = {
  objectMetadataItems: ObjectMetadataItem[];
  objectMetadataItem: Pick<ObjectMetadataItem, 'fields'>;
  depth: 0 | 1;
  shouldOnlyLoadRelationIdentifiers?: boolean;
};

export const generateDepthRecordGqlFieldsFromObject = ({
  objectMetadataItems,
  objectMetadataItem,
  depth,
  shouldOnlyLoadRelationIdentifiers = true,
}: GenerateDepthRecordGqlFields) => {
  return generateDepthRecordGqlFieldsFromFields({
    objectMetadataItems,
    fields: objectMetadataItem.fields,
    depth,
    shouldOnlyLoadRelationIdentifiers,
  });
};
