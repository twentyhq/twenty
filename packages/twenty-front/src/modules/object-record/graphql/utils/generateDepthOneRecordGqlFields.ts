import { ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { isDefined } from 'twenty-shared/utils';

export type GenerateDepthOneRecordGqlFields = {
  objectMetadataItem: ObjectMetadataItem;
};

export const generateDepthOneRecordGqlFields = ({
  objectMetadataItem,
}: GenerateDepthOneRecordGqlFields) =>
  objectMetadataItem.fields.reduce<Record<string, true>>((acc, field) => {
    return {
      ...acc,
      ...(isDefined(field.settings?.joinColumnName)
        ? {
            [field.settings.joinColumnName]: true,
          }
        : {}),
      [field.name]: true,
    };
  }, {});
