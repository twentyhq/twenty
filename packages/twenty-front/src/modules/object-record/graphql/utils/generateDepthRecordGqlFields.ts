import { type ObjectMetadataItem } from '@/object-metadata/types/ObjectMetadataItem';
import { generateDepthRecordGqlFieldsFromFields } from '@/object-record/graphql/utils/generateDepthRecordGqlFieldsFromFields';

export type GenerateDepthRecordGqlFields = {
  objectMetadataItems: ObjectMetadataItem[];
  objectMetadataItem: ObjectMetadataItem;
  depth: 0 | 1;
};

export const generateDepthRecordGqlFields = ({
  objectMetadataItems,
  objectMetadataItem,
  depth,
}: GenerateDepthRecordGqlFields) => {
  return generateDepthRecordGqlFieldsFromFields({
    objectMetadataItems,
    fields: objectMetadataItem.fields,
    depth,
  });
};
