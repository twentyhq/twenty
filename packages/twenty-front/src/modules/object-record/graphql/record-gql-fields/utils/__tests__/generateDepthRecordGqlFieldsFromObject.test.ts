import { generateDepthRecordGqlFieldsFromObject } from '@/object-record/graphql/record-gql-fields/utils/generateDepthRecordGqlFieldsFromObject';
import { generatedMockObjectMetadataItems } from '~/testing/utils/generatedMockObjectMetadataItems';
import { getMockObjectMetadataItemOrThrow } from '~/testing/utils/getMockObjectMetadataItemOrThrow';

describe('generateDepthRecordGqlFieldsFromObject', () => {
  it('should generate depth one record gql fields from object', () => {
    const result = generateDepthRecordGqlFieldsFromObject({
      objectMetadataItem: getMockObjectMetadataItemOrThrow('company'),
      objectMetadataItems: generatedMockObjectMetadataItems,
      depth: 1,
    });

    expect(result).toMatchSnapshot();
  });

  it('should generate depth zero record gql fields from object', () => {
    const result = generateDepthRecordGqlFieldsFromObject({
      objectMetadataItem: getMockObjectMetadataItemOrThrow('company'),
      objectMetadataItems: generatedMockObjectMetadataItems,
      depth: 0,
    });

    expect(result).toMatchSnapshot();
  });
});
