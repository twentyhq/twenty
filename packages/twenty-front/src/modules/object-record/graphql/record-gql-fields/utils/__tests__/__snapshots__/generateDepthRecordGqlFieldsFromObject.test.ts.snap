// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDepthRecordGqlFieldsFromObject should generate depth one record gql fields from object 1`] = `
{
  "accountOwner": {
    "avatarUrl": true,
    "id": true,
    "name": true,
  },
  "accountOwnerId": true,
  "address": true,
  "annualRecurringRevenue": true,
  "attachments": {
    "id": true,
    "name": true,
  },
  "createdAt": true,
  "createdBy": true,
  "deletedAt": true,
  "domainName": true,
  "employees": true,
  "favorites": {
    "id": true,
  },
  "id": true,
  "idealCustomerProfile": true,
  "introVideo": true,
  "linkedinLink": true,
  "name": true,
  "noteTargets": {
    "id": true,
  },
  "opportunities": {
    "id": true,
    "name": true,
  },
  "people": {
    "avatarUrl": true,
    "id": true,
    "name": true,
  },
  "position": true,
  "searchVector": true,
  "tagline": true,
  "taskTargets": {
    "id": true,
  },
  "timelineActivities": {
    "id": true,
  },
  "updatedAt": true,
  "visaSponsorship": true,
  "workPolicy": true,
  "xLink": true,
}
`;

exports[`generateDepthRecordGqlFieldsFromObject should generate depth zero record gql fields from object 1`] = `
{
  "accountOwnerId": true,
  "address": true,
  "annualRecurringRevenue": true,
  "createdAt": true,
  "createdBy": true,
  "deletedAt": true,
  "domainName": true,
  "employees": true,
  "id": true,
  "idealCustomerProfile": true,
  "introVideo": true,
  "linkedinLink": true,
  "name": true,
  "position": true,
  "searchVector": true,
  "tagline": true,
  "updatedAt": true,
  "visaSponsorship": true,
  "workPolicy": true,
  "xLink": true,
}
`;
